hazelcast-tpc-engine/src/test/java/com/hazelcast/internal/tpc/util/CircularQueueTest.java

```diff
@@ -0,0 +1,233 @@
+        new CircularQueue<>(0);
+    }
+
+    @Test
+    public void test_construction() {
+        CircularQueue<Integer> queue = new CircularQueue<>(128);
+        assertTrue(queue.isEmpty());
+        assertFalse(queue.isFull());
+        assertEquals(0, queue.size());
+        assertEquals(128, queue.remaining());
+        assertEquals(128, queue.capacity());
+    }
+
+    @Test
+    public void test_construction_whenCapacityNotPowerOf2() {
+        CircularQueue<Integer> queue = new CircularQueue<>(60);
+        assertEquals(64, queue.capacity());
+    }
+
+    @Test
+    public void test_offer() {
+        CircularQueue<Integer> queue = new CircularQueue<>(128);
+        for (int k = 0; k < queue.capacity() - 1; k++) {
+            assertTrue(queue.offer(k));
+            assertEquals(k + 1, queue.size());
+            assertEquals(queue.capacity() - 1 - k, queue.remaining());
+            assertFalse(queue.isEmpty());
+            assertFalse(queue.isFull());
+        }
+
+        assertTrue(queue.offer(128));
+        assertEquals(queue.capacity(), queue.size());
+        assertEquals(0, queue.remaining());
+        assertFalse(queue.isEmpty());
+        assertTrue(queue.isFull());
+    }
+
+    @Test
+    public void test_offer_whenFull() {
+        CircularQueue<Integer> queue = new CircularQueue<>(128);
+        for (int k = 0; k < queue.capacity(); k++) {
+            queue.offer(k);
+        }
+
+        boolean offered = queue.offer(1234);
+        assertFalse(offered);
+        assertEquals(128, queue.capacity());
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void test_offer_whenNull() {
+        CircularQueue<Integer> queue = new CircularQueue<>(128);
+        queue.offer(null);
+    }
+
+    @Test
+    public void test_peek() {
+        CircularQueue<Integer> queue = new CircularQueue<>(128);
+        assertNull(queue.peek());
+
+        queue.offer(1);
+        assertEquals(Integer.valueOf(1), queue.peek());
+    }
+
+    @Test
+    public void test_add() {
+        CircularQueue<Integer> queue = new CircularQueue<>(128);
+        for (int k = 0; k < queue.capacity() - 1; k++) {
+            queue.add(k);
+            assertEquals(k + 1, queue.size());
+            assertEquals(queue.capacity() - 1 - k, queue.remaining());
+            assertFalse(queue.isEmpty());
+            assertFalse(queue.isFull());
+        }
+
+        queue.add(128);
+        assertEquals(queue.capacity(), queue.size());
+        assertEquals(0, queue.remaining());
+        assertFalse(queue.isEmpty());
+        assertTrue(queue.isFull());
+    }
+
+    @Test
+    public void test_add_whenFull() {
+        CircularQueue<Integer> queue = new CircularQueue<>(128);
+        for (int k = 0; k < queue.capacity(); k++) {
+            queue.add(k);
+        }
+
+        assertThrows(IllegalStateException.class, ()->queue.add(1234));
+    }
+
+    @Test(expected = NullPointerException.class)
```

Comment:
Minor: If we want to use assertThrows for our tests, we can change the tests in this file as well

Reply:
Done